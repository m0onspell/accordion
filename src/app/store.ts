import { configureStore } from '@reduxjs/toolkit';
import settingsSlice from '../features/settings/settingsSlice';
import activitiesSlice from '../features/activities/activitiesSlice';

export const makeStore = () =>
  configureStore({
    reducer: {
      settings: settingsSlice,
      activities: activitiesSlice,
    },
  });

export const store = makeStore();

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type Store = typeof store;
