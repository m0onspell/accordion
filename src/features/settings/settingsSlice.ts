import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../../app/store';

export interface SettingsForm {
  firstName: string;
  lastName: string;
  email: string;
}

export type SettingsState = SettingsForm;

const initialState: SettingsState = {
  firstName: 'John',
  lastName: 'Doe',
  email: 'admin@mail.to',
};

export const settingsSlice = createSlice({
  name: 'settings',
  initialState,
  reducers: {
    save: (state, action: PayloadAction<SettingsState>) => action.payload,
  },
});

export const { save } = settingsSlice.actions;

export const selectSettings = (state: RootState) => state.settings;
export default settingsSlice.reducer;
