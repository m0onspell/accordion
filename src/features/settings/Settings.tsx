import React, { FC, useCallback } from 'react';
import { Formik, Form, Field, ErrorMessage } from 'formik';

import { useAppSelector, useAppDispatch } from '../../app/hooks';
import { save, selectSettings, SettingsForm } from './settingsSlice';
import * as Yup from 'yup';
import styles from './Settings.module.css';

const SettingsSchema = Yup.object().shape({
  firstName: Yup.string().min(2, 'Too Short!').max(50, 'Too Long!').required('Required'),
  lastName: Yup.string().min(2, 'Too Short!').max(50, 'Too Long!').required('Required'),
  email: Yup.string().email('Invalid email').required('Required'),
});

const Settings: FC = () => {
  const settings = useAppSelector(selectSettings);
  const dispatch = useAppDispatch();
  const onSubmit = useCallback(
    values => {
      dispatch(save(values));
    },
    [dispatch],
  );
  return (
    <>
      <h1>Settings</h1>
      <Formik<SettingsForm>
        initialValues={settings}
        validationSchema={SettingsSchema}
        onSubmit={onSubmit}
      >
        <Form>
          <div className={styles.formGroup}>
            <label htmlFor="firstName">First Name</label>
            <Field id="firstName" name="firstName" placeholder="First Name" />
            <div className={styles.errorMessage}>
              <ErrorMessage className={styles.errorMessage} name="firstName" />
            </div>
          </div>

          <div className={styles.formGroup}>
            <label htmlFor="lastName">Last Name</label>
            <Field id="lastName" name="lastName" placeholder="Last Name" />
            <div className={styles.errorMessage}>
              <ErrorMessage className={styles.errorMessage} name="lastName" />
            </div>
          </div>

          <div className={styles.formGroup}>
            <label htmlFor="email">Email</label>
            <Field id="email" name="email" placeholder="Email" />
            <div className={styles.errorMessage}>
              <ErrorMessage name="email" />
            </div>
          </div>

          <button className={styles.submitButton} type="submit">
            Submit
          </button>
        </Form>
      </Formik>
    </>
  );
};

export { Settings };
