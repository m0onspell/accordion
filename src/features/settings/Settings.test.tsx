import React from 'react';
import renderer from 'react-test-renderer';
import { Settings } from './Settings';
import { Provider } from 'react-redux';
import { makeStore } from '../../app/store';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { selectSettings } from './settingsSlice';

const setup = (store = makeStore()) => (
  <Provider store={store}>
    <Settings />
  </Provider>
);

test('renders correctly', async () => {
  const tree = renderer.create(setup());
  expect(tree.toJSON()).toMatchSnapshot();
});

test('saves changed settings to the store', async () => {
  const store = makeStore();

  render(setup(store));

  const firstName = screen.getByLabelText(/First Name/i);
  const lastName = screen.getByLabelText(/Last Name/i);
  const email = screen.getByLabelText(/Email/i);
  const submit = screen.getByRole('button', { name: /submit/i });

  userEvent.clear(firstName);
  userEvent.type(firstName, 'Ben');

  userEvent.clear(lastName);
  userEvent.type(lastName, 'Dukes');

  userEvent.clear(email);
  userEvent.type(email, 'newmail@to.com');

  userEvent.click(submit);

  await waitFor(() =>
    expect(selectSettings(store.getState())).toEqual({
      email: 'newmail@to.com',
      firstName: 'Ben',
      lastName: 'Dukes',
    }),
  );
});

test('adds required error', () => {
  render(setup());

  const inputs = [
    screen.getByLabelText(/First Name/i),
    screen.getByLabelText(/Last Name/i),
    screen.getByLabelText(/Email/i),
  ];
  const submit = screen.getByRole('button', { name: /submit/i });

  inputs.forEach(async input => {
    userEvent.clear(input);
    userEvent.click(submit);
    await waitFor(() =>
      // @ts-ignore
      expect(input.parentElement.querySelector('.errorMessage')).toHaveTextContent('Required!'),
    );
  });
});

test('adds email error', async () => {
  render(setup());

  const email = screen.getByLabelText(/Email/i);
  const submit = screen.getByRole('button', { name: /submit/i });

  userEvent.clear(email);
  userEvent.type(email, '123');
  userEvent.click(submit);

  await waitFor(() =>
    // @ts-ignore
    expect(email.parentElement.querySelector('.errorMessage')).toHaveTextContent('Invalid email'),
  );
});
