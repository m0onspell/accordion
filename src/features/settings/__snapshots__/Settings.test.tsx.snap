// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <h1>
    Settings
  </h1>,
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="formGroup"
    >
      <label
        htmlFor="firstName"
      >
        First Name
      </label>
      <input
        id="firstName"
        name="firstName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="First Name"
        value="John"
      />
      <div
        className="errorMessage"
      />
    </div>
    <div
      className="formGroup"
    >
      <label
        htmlFor="lastName"
      >
        Last Name
      </label>
      <input
        id="lastName"
        name="lastName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Last Name"
        value="Doe"
      />
      <div
        className="errorMessage"
      />
    </div>
    <div
      className="formGroup"
    >
      <label
        htmlFor="email"
      >
        Email
      </label>
      <input
        id="email"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Email"
        value="admin@mail.to"
      />
      <div
        className="errorMessage"
      />
    </div>
    <button
      className="submitButton"
      type="submit"
    >
      Submit
    </button>
  </form>,
]
`;
